File: backtracking/78.subsets.ts
// @leet start
function subsets(nums: number[]): number[][] {
  const res: number[][] = [];
  const path: number[] = [];
  function dfs(i: number) {
    res.push([...path]);
    for (let j = i; j < nums.length; j++) {
      path.push(nums[j]);
      dfs(j + 1);
      path.pop();
    }
  }
  dfs(0);
  return res;
}
// @leet end
function subsets2(nums: number[]): number[][] {
  const res: number[][] = [];
  const path: number[] = [];
  function dfs(i: number) {
    if (i === nums.length) {
      res.push([...path]);
      return;
    }
    dfs(i + 1);
    path.push(nums[i]);
    dfs(i + 1);
    path.pop();
  }
  dfs(0);
  return res;
}

File: Backtracking/78.subsets.py
from typing import List


# @leet start
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res, path, n = [], [], len(nums)

        def dfs(idx):
            res.append(path[:])
            for i in range(idx, n):
                path.append(nums[i])
                dfs(i + 1)
                path.pop()

        dfs(0)
        return res


# @leet end
class Solution2:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res, path, n = [], [], len(nums)

        def dfs(idx):
            if idx == n:
                res.append(path[:])
                return
            dfs(idx + 1)
            path.append(nums[idx])
            dfs(idx + 1)
            path.pop()

        dfs(0)
        return res


